src/api.rs

```diff
@@ -1,33 +1,151 @@
+
+// Struct for req to `/api/{game_id}/submit-move`
+// This is basically the game::adaptor::GenericGameMove struct
+
+/// Struct for res from `/api/{game_id}/submit-move`
+#[derive(Serialize)]
+pub struct SubmitMoveRes {
+    success: bool,
+}
+
+/// Struct for res from `/api/{game_id}/submit-move`
+#[derive(Serialize)]
+pub struct WaitForUpdateRes {
+    updated: bool
+}
 
 #[post("/api/create-game")]
-pub(crate) async fn create_game(mut payload: Payload) -> impl Responder {
-    todo!(); "not yet implemented"
+pub(crate) async fn create_game(mut payload: web::Json<CreateGameReq>,
+                                gm_wrapped: web::Data<RwLock<GameManager>>) -> Result<Json<CreateGameRes>> {
+    // Validate & Pass it on
+    let mut gm = gm_wrapped.write().unwrap();
+    let game_id = match payload.name.as_str() {
+        "connect_4" => gm.create_game(|id| Box::new(connect4::Connect4Adapter::new(id))),
+        _ => return Err(Error::from(NoSuchGameError(payload.name.clone()))) // TODO: Should I return an empty string or an error
+    };
+    game_id.and_then(|id| Ok(Json(CreateGameRes { game_id: id.to_string()[4..].parse()? })))
 }
 
 #[get("/api/list-games")]
-pub(crate) async fn list_games(mut payload: Payload) -> impl Responder {
-    todo!(); "not yet implemented"
+pub(crate) async fn list_games() -> Result<Json<ListGamesRes>> {
+    todo!()
 }
 
 #[post("/api/join-game")]
-pub(crate) async fn join_game(mut payload: Payload) -> impl Responder {
-    todo!(); "not yet implemented"
+pub(crate) async fn join_game(mut payload: web::Json<JoinGameReq>,
+                              gm_wrapped: web::Data<RwLock<GameManager>>) -> Result<Json<JoinGameRes>> {
+    // Validate input
+    let game_id = validate_id(&payload.game_id)?;
+    // Join a game
+    let mut gm = gm_wrapped.write().unwrap();
+    let session_id = gm.receive_join(GameId::from(game_id), payload.username.clone());
+    todo!()
 }
 
 #[get("/api/{game_id}/get-state")]
-pub(crate) async fn get_state(mut payload: Payload) -> impl Responder {
-    todo!(); "not yet implemented"
+pub(crate) async fn get_state(web::Path((game_id)): web::Path<String>, gm_wrapped: web::Data<RwLock<GameManager>>) -> Result<Json<adapter::GenericGameState>> {
+    // Validation & get_state
+    let game_id = validate_id(&game_id)?;
+    let gm = gm_wrapped.read().unwrap();
+    // Get state and return
+    Ok(Json(gm.get_state(GameId::from(game_id))?))
 }
 
 #[post("/api/{game_id}/submit-move")]
-pub(crate) async fn submit_move(mut payload: Payload) -> impl Responder {
-    todo!(); "not yet implemented"
+pub(crate) async fn submit_move(web::Path((game_id)): web::Path<String>,
+                                mut payload: web::Json<adapter::GenericGameMove>,
+                                gm_wrapped: web::Data<RwLock<GameManager>>) -> Result<Json<SubmitMoveRes>> {
+    let game_id = validate_id(&game_id)?;
+    // Submit that to game manager
+    let mut gm = gm_wrapped.write().unwrap();
```

Comment:
We shouldn't need a write lock to submit a move as the game adapter map uses interior mutability through `Mutex`

Reply:
Ok